### PR TITLE
Implement SendGrid service for featured Author Emails

### PR DESCRIPTION
**Pull request by davidnso**
This PR implements a SendGrid service that can be injected anywhere else on the server for ease of use so that we can expand as things grow and change.

### CODE REVIEW

File: src/learning-objects/learning-objects.controller.ts [hunk 1]
```diff
@@ -17,15 +18,28 @@ export class LearningObjectsController {
...
     @Patch()
     @HttpCode(201)
     async updateAll(@Body() learningObjectDto: LearningObjectDto[]): Promise<{ message: string }> { 
         if (learningObjectDto.length != 5)
         {
             throw new BadRequestException('ERROR!! Array must contain 5 Learning Objects, nothing more & nothing less!!')
         }
-        await this.learningObjectService.updateAllFeatured(learningObjectDto);
+       
+        // await this.learningObjectService.updateAllFeatured(learningObjectDto);
```
Comment (pzalep1) [inline src/learning-objects/learning-objects.controller.ts:30]: :fire: Comment

File: src/services/send-grid/send-grid.service.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      objectName: string,
+  }[]){ 
+      try {
+        for(let info of mailinfo){ 
+            const mailDefauls = { 
+                to: info.email,
+                ...MAILER_DEFAULTS,
+                subject: 'Congratulations You\ve been featured on Clark!!',
+                templateId: SENDGRID_TEMPLATES.FEATURED_AUTHOR,
+                templateVars: {
```
Comment (pzalep1) [inline src/services/send-grid/send-grid.service.ts:31]: Depending on where @EVogel1999 created the template these could be dynamic_template_data. Legacy SendGrid templates still use this, but new dynamic templates have a different variable name.

File: src/services/send-grid/send-grid.service.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                    object: { 
+                        name: info.objectName,
+                        cuid: info.cuid
+                    }
+                }
+            }
+            await this.sgclient.send(mailDefauls);
+         } 
+      } catch (error) {
+        console.log(error.response.body.errors)
```
Comment (pzalep1) [inline src/services/send-grid/send-grid.service.ts:45]: Can we throw this so an error gets reported back?

File: src/services/send-grid/interfaces/mailer.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+export interface DefaultEmailParams {
```
Comment (pzalep1) [inline src/services/send-grid/interfaces/mailer.ts:1]: Can we also create enums for the templates so we have a place to put template ids? Like here: https://github.com/Cyber4All/cards-service/blob/master/src/shared/interfaces/mailer.defaults.ts

**Review by pzalep1 — CHANGES_REQUESTED** [review]

File: src/services/send-grid/send-grid.service.ts (hunk 3 above)
Comment (davidnso) [reply src/services/send-grid/send-grid.service.ts:45]: I'm pretty sure the errors are handled within the service. This was just for easier debugging. It's a little weird

Comment (pzalep1) [reply src/services/send-grid/send-grid.service.ts:45]: Okay that sounds fine. Nest handles a lot of these errors, but I wasn’t sure if it was meant to be thrown somewhere.

**Comment by Olaniyan99** [comment]
apart from the comment Paige pointed out, this is good to go.

File: src/learning-objects/learning-objects.controller.ts [hunk 5]
```diff
@@ -17,15 +18,28 @@ export class LearningObjectsController {
         return this.learningObjectService.getOneLearningObject(cuid);
     }
 
-    @UseGuards(JwtAuthGuard)
+    // @UseGuards(JwtAuthGuard)
```
Comment (EVogel1999) [inline src/learning-objects/learning-objects.controller.ts:21]: Why is this commented out?

File: src/services/send-grid/send-grid.service.ts (hunk 2 above)
Comment (EVogel1999) [reply src/services/send-grid/send-grid.service.ts:31]: I created it under dynamic templates.  The data needed is located in the design/implementation story for the email.

File: src/learning-objects/learning-objects.controller.ts (hunk 5 above)
Comment (davidnso) [reply src/learning-objects/learning-objects.controller.ts:21]: I needed to bypass the authgraud. This is uncommented the next commit

File: src/services/send-grid/interfaces/mailer.defaults.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+export enum TEMPLATES {
+    FEATURED_AUTHOR = 'featured author',
+}
+
+export const TEMPLATE_VARIABLES = new Map<
+    TEMPLATES,
+    { [index: string]: any }
+>();
+
+TEMPLATE_VARIABLES.set(TEMPLATES.FEATURED_AUTHOR, {
```
Comment (pzalep1) [inline src/services/send-grid/interfaces/mailer.defaults.ts:17]: I like this.

**Review by pzalep1 — APPROVED** [review]

**Review by EVogel1999 — APPROVED** [review]

**Pull request merged**